File: benchmarking/benchmarks/MNIST/python/mnist.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from keras.layers import Input, Convolution2D, MaxPooling2D, Activation, Dropout, Reshape, Dense
from keras import optimizers
from keras.models import Model

import tensorflow as tf
import dataset

"""""""""""
DEFAULTS
"""""""""""
TRAIN_SIZE = 1000
TEST_SIZE = 1000
LEARNING_RATE = .001
BATCH_SIZE = 64
IMAGE_SIZE = 28
IMAGE_DEPTH = 1
EPOCHS = 1


class MNIST:
    def __init__(self, backend):

        self.train_data = dataset.train("/tmp/mnist_data")
        self.test_data = dataset.test("/tmp/mnist_data")
        self.backend = backend

        self.train_images = self.train_data['images'][:TRAIN_SIZE]
        self.train_labels = self.train_data['labels'][:TRAIN_SIZE]

        self.test_images = self.test_data['images'][:TRAIN_SIZE]
        self.test_labels = self.test_data['labels'][:TRAIN_SIZE]

        if backend == 'gpu':
            self.device = "/device:GPU:0"
        else:
            assert backend == 'cpu', 'Invalid backend specified: %s' % backend
            self.device = "/cpu:0"

        self.model = MNIST.create_model()

    @staticmethod
    def create_model(input_shape=(IMAGE_SIZE, IMAGE_SIZE, IMAGE_DEPTH)):
        img_input = Input(shape=input_shape)

        x = Convolution2D(32, (5, 5), strides=(1, 1), padding='same', name='conv1')(img_input)
        x = Activation('relu', name='relu_conv1')(x)
        x = MaxPooling2D(pool_size=(2, 2), strides=(2, 2), name='pool1')(x)

        x = Convolution2D(64, (5, 5), strides=(1, 1), padding='same', name='conv2')(x)
        x = Activation('relu', name='relu_conv2')(x)
        x = MaxPooling2D(pool_size=(2, 2), strides=(2, 2), name='pool2')(x)

        x = Reshape([-1, 7 * 7 * 64])(x)
        x = Dense(1024)(x)
        x = Activation('relu', name='relu_conv1')(x)
        x = Dropout(0.4)(x)
        x = Dense(10)(x)

        inputs = img_input

        model = Model(inputs, x, name='squeezenet')

        optimizer = optimizers.adam(lr=0.001)
        model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
        return model

    def train(self, train_steps=EPOCHS):
        with tf.device(self.device):
            self.model.fit(self.train_images, self.train_labels, batch_size=BATCH_SIZE, epochs=train_steps)

    def predict(self):
        with tf.device(self.device):
            self.model.evaluate(x=self.test_images, y=self.test_labels, batch_size=BATCH_SIZE)


def init(backend):
    return MNIST(backend)

File: benchmarking/benchmarks/MNIST/javascript/mnist.js
import * as tf from '@tensorflow/tfjs';
import {MnistData} from "./data";

/*****************************
 *  CONSTANTS
 ****************************/
// Data
const d = new MnistData();

// Hyper-parameters
const LEARNING_RATE = .001;
const BATCH_SIZE = 64;
const EPOCHS = 1;

// Data constants.
const IMAGE_SIZE = 28;
const IMAGE_DEPTH = 1;
const LABELS_SIZE = 10;
const TRAIN_SIZE = 10000;
const TEST_SIZE = 10000;

export class MNIST {

    constructor(){
        this.model = MNIST.create_model();
    }

    // Create the model and assigns it to the global model property
    static create_model() {
        const optimizer = tf.train.adam(LEARNING_RATE);
        const model = tf.sequential();

        model.add(tf.layers.conv2d({
            inputShape: [IMAGE_SIZE, IMAGE_SIZE, IMAGE_DEPTH],
            kernelSize: 5,
            filters: 32,
            strides: 1,
            activation: 'relu',
            kernelInitializer: 'randomNormal',
            biasInitializer: 'zeros'
        }));

        model.add(tf.layers.maxPooling2d({poolSize: [2, 2], strides: [2, 2]}));

        model.add(tf.layers.conv2d({
            kernelSize: 5,
            filters: 64,
            strides: 1,
            activation: 'relu',
            kernelInitializer: 'randomNormal',
            biasInitializer: 'zeros'
        }));

        model.add(tf.layers.maxPooling2d({poolSize: [2, 2], strides: [2, 2]}));

        model.add(tf.layers.flatten());

        model.add(tf.layers.dense({
            units: 1024, kernelInitializer: 'randomNormal',
            biasInitializer: 'zeros', activation: 'softmax'
        }));

        model.add(tf.layers.dropout({
            rate: 0.4
        }));

        model.add(tf.layers.dense({
            units: LABELS_SIZE, kernelInitializer: 'randomNormal',
            biasInitializer: 'zeros', activation: 'softmax'
        }));

        model.compile({
            optimizer: optimizer,
            loss: 'categoricalCrossentropy',
            metrics: ['accuracy'],
        });

        return model
    }

    // Train the model.
    async train(batch, epochs = EPOCHS) {
        await this.model.fit(batch.images.reshape([TRAIN_SIZE, IMAGE_SIZE, IMAGE_SIZE, IMAGE_DEPTH]), batch.labels,
            {batchSize: BATCH_SIZE, epochs: epochs});
    }

    // Predict the digit number from a batch of input images.
    async predict(batch){
        await this.model.predict(batch.images.reshape([TEST_SIZE, IMAGE_SIZE, IMAGE_SIZE, IMAGE_DEPTH]),
            {batchSize: BATCH_SIZE});
    }
}


/*****************************
 * HELPERS
 ****************************/
// Gets the next shuffled training batch
export function nextBatch(type) {
    return (type === 'train') ? d.nextTrainBatch(TRAIN_SIZE) : d.nextTestBatch(TEST_SIZE);
}

// Sets the data in the data.js file
async function set_data(){
    await d.load()
}

/*****************************
 *  SETUP
 ****************************/
export async function init(backend) {

    console.log(`Info: Current backend is ${tf.getBackend()}`);

    // Set backend to run on either CPU or GPU
    if(backend === 'gpu' || backend === 'cpu'){
        (backend === 'gpu') ? tf.setBackend('webgl') : tf.setBackend('cpu');
    } else {
        throw new Error(`Invalid backend parameter: ${backend}. Please specify either 'cpu' or 'gpu'`)
    }

    console.log("Info: Set backend");

    await set_data();
    return new MNIST()
}
